maketab.c

```diff
@@ -136,8 +136,12 @@ int main(int argc, char *argv[])
 	while (fgets(buf, sizeof buf, fp) != NULL) {
 		// 199 is sizeof(def) - 1
 		n = sscanf(buf, "%1c %199s %199s %d", &c, def, name, &tok);
-		if (c != '#' || (n != 4 && strcmp(def,"define") != 0))	/* not a valid #define */
-			continue;
+		if (c != '#' || (n != 4 && strcmp(def, "define") != 0)) {
+			/* not a valid #define, bison uses enums now */
+			n = sscanf(buf, "%199s = %d,\n", name, &tok);
+			if (n != 2)
+				continue;
+		}
```

Comment:
This change breaks things, badly. When I check out your branch on my Ubuntu 16.04 system, I get:
```sh
gcc -g -Wall -pedantic -Wcast-qual -O2 maketab.c -o maketab
./maketab ytab.h >proctab.c
gcc -g -Wall -pedantic -Wcast-qual   -O2   -c -o proctab.o proctab.c
proctab.c:101:2: warning: excess elements in array initializer
  "FIRSTTOKEN", /* 258 */
  ^
proctab.c:101:2: note: (near initialization for ‘printname’)
proctab.c:102:2: warning: excess elements in array initializer
  "PROGRAM", /* 259 */
  ^
```
And lots more of the same.  Please revert or fix.

Reply:
Yes, that's because of -y (this is not needed now). I will explain in the commit message.

Reply:
Bison is indeed happier now. However, testing with `byacc`, I get the following:
```sh
./maketab ytab.h >proctab.c
gcc -g -Wall -pedantic -Wcast-qual   -O2   -c -o proctab.o proctab.c
proctab.c:5:43: warning: ISO C forbids empty initializer braces [-Wpedantic]
 static const char * const printname[95] = {
```
Here are the first few lines of ytab.h:
```
$ head ytab.h
#define FIRSTTOKEN 257
#define PROGRAM 258
#define PASTAT 259
#define PASTAT2 260
```

Reply:
should be all fixed now. Tested with:
nbyacc -d
bison -d -y
bison -d

Reply:
So, there is still a problem with bison -d:
```sh
$ git diff
diff --git a/makefile b/makefile
index 735436c..ea749d3 100644
--- a/makefile
+++ b/makefile
@@ -35,7 +35,7 @@ CC = $(HOSTCC)  # change this is cross-compiling.
 
 # yacc options.  pick one; this varies a lot by system.
 #YFLAGS = -d -S
-YACC = bison -d -y
+YACC = bison -d
 #YACC = yacc -d
 #              -S uses sprintf in yacc parser instead of sprint
 
$ make
make: Circular ytab.c <- ytab.c dependency dropped.
bison -d  awkgram.y
awkgram.y: warning: 44 shift/reduce conflicts [-Wconflicts-sr]
awkgram.y: warning: 85 reduce/reduce conflicts [-Wconflicts-rr]
mv y.tab.c ytab.c
mv: cannot stat 'y.tab.c': No such file or directory
makefile:64: recipe for target 'ytab.c' failed
make: *** [ytab.c] Error 1
```
The expected filenames are not created.

Reply:
So the simple fix here is to just:

--- a/makefile
+++ b/makefile
@@ -62,8 +62,10 @@ $(OFILES):   awk.h ytab.h proto.h
 #each others outputs.
 y%.c y%.h:     awk.h proto.h awkgram.y
        $(YACC) $(YFLAGS) awkgram.y
-       mv y.$*.c y$*.c
-       mv y.$*.h y$*.h
+       if [ -f y.$*.c ]; then mv y.$*.c ytab.c; fi
+       if [ -f y.$*.h ]; then mv y.$*.h ytab.h; fi
+       if [ -f awkgram.$*.c ]; then mv awkgram.$*.c ytab.c; fi
+       if [ -f awkgram.$*.h ]; then mv awkgram.$*.h ytab.h; fi
 
But the new makefile is not portable since it uses pattern rules which for example bsdmake does not have. It also complains about a cycle through ytab.c

Perhaps we should at this point use "-b awkgram" in both yacc and bison invocations so that they create awkgram.tab.[ch] so that they are consistent and then we don't need to use pattern rules... Of course this is not portable for 20+ year old yacc implementations.

Please let me know how you want to proceed.

Reply:
Instead of using pattern rules, just hard code the dependencies:
```
y.tab.c y.tab.h: awkgram.y
    $(YACC) ...
    if ....
```
I want maximal portability and to keep things simple. Thanks.